activerecord/test/cases/attribute_methods_test.rb

```diff
@@ -1246,6 +1246,92 @@ class ChildWithDeprecatedBehaviorResolved < ClassWithDeprecatedAliasAttributeBeh
     assert_equal("overridden_subject_was", obj.subject_was)
   end
 
+  ClassWithGeneratedAttributeMethodTarget = Class.new(ActiveRecord::Base) do
+    self.table_name = "topics"
+    alias_attribute :saved_title, :title_in_database
+  end
+
+  test "#alias_attribute with an _in_database method issues a deprecation warning" do
+    message = <<~MESSAGE.gsub("\n", " ")
+    AttributeMethodsTest::ClassWithGeneratedAttributeMethodTarget model aliases `title_in_database`, but title_in_database is not an attribute.
+    Starting in Rails 7.2 `, alias_attribute with non-attribute targets will raise. Use `alias_method :saved_title`, :title_in_database or define the method manually.
+    MESSAGE
+
+    obj = assert_deprecated(message, ActiveRecord.deprecator) do
+      ClassWithGeneratedAttributeMethodTarget.new
+    end
+    obj.title = "A river runs through it"
+    assert_nil obj.saved_title
+    obj.save
+    assert_equal "A river runs through it", obj.saved_title
+  end
+
+  ClassWithEnumMethodTarget = Class.new(ActiveRecord::Base) do
+    self.table_name = "books"
+
+    attribute :status, :string
+    enum status: {
+      pending: "0",
+      completed: "1",
+    }
+    alias_attribute :is_pending?, :pending?
+  end
+
+  test "#alias_attribute with enum method issues a deprecation warning" do
+    message = <<~MESSAGE.gsub("\n", " ")
+    AttributeMethodsTest::ClassWithEnumMethodTarget model aliases `pending?`, but pending? is not an attribute.
+    Starting in Rails 7.2 `, alias_attribute with non-attribute targets will raise.
+    Use `alias_method :is_pending?`, :pending? or define the method manually.
+    MESSAGE
+
+    obj = assert_deprecated(message, ActiveRecord.deprecator) do
+      ClassWithEnumMethodTarget.new
+    end
+    obj.status = "pending"
+    assert obj.pending?
+    assert obj.is_pending?
+  end
+
+  ClassWithAssociationTarget = Class.new(ActiveRecord::Base) do
+    self.table_name = "books"
+    belongs_to :author
+
+    alias_attribute :written_by, :author
+  end
+
+  test "#alias_attribute with an association method issues a deprecation warning" do
+    message = <<~MESSAGE.gsub("\n", " ")
+    AttributeMethodsTest::ClassWithAssociationTarget model aliases `author`, but author is not an attribute.
+    Starting in Rails 7.2 `, alias_attribute with non-attribute targets will raise.
+    Use `alias_method :written_by`, :author or define the method manually.
+    MESSAGE
+
+    assert_deprecated(message, ActiveRecord.deprecator) do
+      ClassWithAssociationTarget.new
```

Comment:
Do we want to add an assertion that `.written_by` still works? It's not what this PR is concerned about but would be nice to have additional coverage for the whole `alias_attribute` set of changes

Reply:
Good call, updated as part of [this change](https://github.com/rails/rails/compare/02eaf46709ceabb9eabcee6ed995439ef632f7c0..22f8548bea2127222fddaf84b2488c4b93661742)